### ISSUE TITLE
InputDecorator should have more specifiy return type than "any"

### ISSUE THREAD

**Issue by Humi171421** [angular/angular#27595]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

 No

### Description
`InputDecorator`, when called as a function, has the return type `any`.
Using `@Input()` therefore causes a violation of TSLint's no-unsafe-any rule.

## 🔬 Minimal Reproduction
Use `@Input()` and run TSLint with rule "no-unsafe-any" set to true

`MyComponent.ts`
<pre><code>
export class MyComponent {
	@Input() public someInput: string;
}
</code></pre>

`Tslint.json`
<pre><code>no-unsafe-any: true</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>TSLint: 5.11.0
Typescript: 3.1.3
</code></pre>

**Comment by Melmoth-the-Wanderer** [comment]
This is similar to #26942 and comes from https://github.com/palantir/tslint/issues/3784.

**Comment by snebjorn** [comment]
#26942 was closed but @Input and @Output decorators still have this issue

**Comment by SchnWalter** [comment]
Actually, there are multiple decorators that have the same issues: `@Input()`, `@Output()`, `@HostBinding()` and `@HostListener()`.

And I've tried updating the types, but it's a bit tricky, since we need to be able to support normal properties but also setters and getters and I couldn't figure out what's the proper return type for all these cases.

**Comment by dudewad** [comment]
Yeah this is really strange behavior that I'm only seeing in the last couple days. It seems very odd that any decorators would be firing errors like this. I imagine it will lead a lot of people to shut off linting or write ugly tslint disable comments :(

**Comment by geremora** [comment]
Yes. Using Angular 9.1, typescript 3.8 and tslint 6.0.0-beta0 is still happening.

**Comment by rafaelss95** [comment]
@kara Can someone from community send a PR for this? I'm interested in working on this.

<hr>

Just to be sure about the changes, taking the `@Input` as an example...

this:

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/core/src/metadata/directives.ts#L688

should be `(bindingPropertyName?: string): TypeDecorator;`?

and 

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/core/src/metadata/directives.ts#L689

should be `new(bindingPropertyName?: string): Input;`?
